### PR TITLE
Use parameters when executing prepared statements rather than fmt.Sprintf

### PR DESCRIPTION
**Pull request by Valarissa**
We should, when possible, use the built-in parameters and `?` format when preparing and executing a query. This is done to prevent SQL Injection attacks.

### CODE REVIEW

**Comment by ncabatoff** [comment]
I approve of the spirit, but I'm not sure all these changes will work.  Have you tested them?  Many RDBMs only support placeholders in certain parts of queries.  I'm not familiar enough with MSSQL to say whether these are problematic, but the fact that there's no automated tests (AFAIK) makes me very uncomfortable.

One example: ALTER LOGIN takes an unquoted login as its first parameter.  Generally placeholders are expanded as quoted strings, which at least in some RDBMs means they can't be used in this context.

**Comment by Valarissa** [comment]
I have not tested them thoroughly, as I don't have mssql on my machine. I was hoping the automated tests would cover it, but I guess my assumptions there may have been wrong? I can go through and change any queries that we suspect won't work. I leaned a bit more towards the security side based on the ticket I was trying to address.

**Comment by Valarissa** [comment]
@ncabatoff I've updated this based on results I was getting from running the tests found in `plugins/database/mssql` and `builtin/logical/mssql` Though, amusingly, running the tests in `plugins/database/mssql` changes your root credentials and doesn't change them back, so all tests after that one fail.

Anyways, these changes have been verified now.

I also updated the go MSSQL driver package.

File: builtin/logical/mssql/secret_creds.go [hunk 1]
```diff
@@ -105,13 +104,13 @@ func (b *backend) secretCredsRevoke(ctx context.Context, req *logical.Request, d
 	// we need to drop the database users before we can drop the login and the role
 	// This isn't done in a transaction because even if we fail along the way,
 	// we want to remove as much access as possible
-	stmt, err := db.Prepare(fmt.Sprintf("EXEC master.dbo.sp_msloginmappings '%s';", username))
+	stmt, err := db.Prepare("EXEC master.dbo.sp_msloginmappings '?';")
```
Comment (ncabatoff) [inline builtin/logical/mssql/secret_creds.go]: I feel like (and the other EXEC below) might've worked if you'd used `?` without quotes.  Can you try it that way please?

Comment (Valarissa) [reply builtin/logical/mssql/secret_creds.go]: Sure thing!

Comment (Valarissa) [reply builtin/logical/mssql/secret_creds.go]: Unfortunately, it doesn't actually work. :/

Comment (ncabatoff) [reply builtin/logical/mssql/secret_creds.go]: I've been reading the microsoft docs and googling and I can't see any reason why this wouldn't work, or why EXEC should be special.  How does it fail?

Comment (Valarissa) [reply builtin/logical/mssql/secret_creds.go]: Well, take your pick:
 - with `'?'`: `"failed to revoke entry: resp: (*logical.Response)(nil) err: mssql: '?' is not a valid login or you do not have permission."`
 - with `?`: `"failed to revoke entry: resp: (*logical.Response)(nil) err: mssql: Incorrect syntax near '?'."`

I think what's happening is that prepare doesn't know how to work with stored procedures.

**Comment by ncabatoff** [comment]
This looks great, I have two last requests:
- remove any remaining use of `?` as placeholder, I think we should consistently use whatever placeholder format a driver prefers (`@x` in this case)
- instead of having a duplicate private `prepareTestContainer`, factor those out into a shared public helper under vault/helper/testhelpers

**Pull request merged**